src/actions/measurements.ts

```diff
@@ -203,63 +264,92 @@ const parseMeasurementsJSON = (json) => {
-          const filterObject = collection.filters.get(field) || { values: new Set()};
-          filterObject.values.add(fieldValue);
-          collection.filters.set(field, filterObject);
-        }
+          // Only save the unique values if the field is in defined filters
+          // OR there are no JSON defined filters, so all fields are filters
+          if ((collection.filters.has(field)) || !collectionFiltersArray) {
+            const filterObject = collection.filters.get(field) || { values: new Set()};
+            filterObject.values.add(fieldValueString);
+            collection.filters.set(field, filterObject);
+          }
 
-        // Save grouping field values and counts
-        if (field in groupingsValues) {
-          const previousValue = groupingsValues[field][fieldValue];
-          groupingsValues[field][fieldValue] = previousValue ? previousValue + 1 : 1;
+          // Save grouping field values and counts
+          if (field in groupingsValues) {
+            const previousValue = groupingsValues[field].get(fieldValueString);
+            groupingsValues[field].set(fieldValueString, previousValue ? previousValue + 1 : 1);
+          }
         }
       });
 
-      // Add stable id for each measurement to help visualization
-      measurement[measurementIdSymbol] = index;
+
+      if (isMeasurement(parsedMeasurement)) {
+        return parsedMeasurement;
+      } else {
+        throw new Error("Could not parse a valid measurement");
+      }
     });
 
     // Create groupings Map for easier access of sorted values and to keep groupings ordering
     // Must be done after looping through measurements to build `groupingsValues` object
     collection.groupings = new Map(
-      collection.groupings.map(({key, order}) => {
-        const valuesByCount = Object.entries(groupingsValues[key])
+      jsonCollection.groupings.map(({key, order}) => {
+        const defaultOrder = order ? order.map((x) => x.toString()) : [];
+        const valuesByCount = [...groupingsValues[key].entries()]
           // Use the grouping values' counts to sort the values, highest count first
           .sort(([, valueCountA], [, valueCountB]) => valueCountB - valueCountA)
           // Filter out values that already exist in provided order from JSON
-          .filter(([fieldValue]) => order ? !order.includes(fieldValue) : true)
+          .filter(([fieldValue]) => !defaultOrder.includes(fieldValue))
           // Create array of field values
           .map(([fieldValue]) => fieldValue);
+
         return [
           key,
           // Prioritize the provided values order then list values by count
-          {values: (order || []).concat(valuesByCount)}
+          {values: (defaultOrder).concat(valuesByCount)}
         ];
       })
     );
+
+    if (isCollection(collection)) {
+      return collection;
+    } else {
+      throw new Error("Could not parse a valid collection");
+    }
```

Comment:
If following the [suggestion](https://github.com/nextstrain/auspice/pull/1910#discussion_r1848966360) to have a pair of narrowing functions `isCollection`+`asCollection`, this can be replaced by `return asCollection(collection)`.

Reply:
Updated in https://github.com/nextstrain/auspice/pull/1910/commits/9a60742f2b51eedd2e3544dad494367679fc88ce